plugin/trino-hive/src/main/java/io/trino/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -223,22 +230,37 @@ public BackgroundHiveSplitLoader(
         this.recursiveDirWalkerEnabled = recursiveDirWalkerEnabled;
         this.ignoreAbsentPartitions = ignoreAbsentPartitions;
         this.optimizeSymlinkListing = optimizeSymlinkListing;
+        requireNonNull(executor, "executor is null");
+        // direct executor is not supported in this implementation due to locking specifics
+        checkExecutorIsNotDirectExecutor(executor);
         this.executor = executor;
         this.partitions = new ConcurrentLazyQueue<>(partitions);
         this.hdfsContext = new HdfsContext(session);
         this.validWriteIds = requireNonNull(validWriteIds, "validWriteIds is null");
         this.maxSplitFileSize = requireNonNull(maxSplitFileSize, "maxSplitFileSize is null");
+        this.maxPartitions = maxPartitions;
     }
 
     @Override
     public void start(HiveSplitSource splitSource)
     {
         this.hiveSplitSource = splitSource;
         this.stopwatch = Stopwatch.createStarted();
-        for (int i = 0; i < loaderConcurrency; i++) {
-            ListenableFuture<Void> future = ResumableTasks.submit(executor, new HiveSplitLoaderTask());
-            addExceptionCallback(future, hiveSplitSource::fail); // best effort; hiveSplitSource could be already completed
+        addLoaderIfNecessary();
+    }
+
+    private void addLoaderIfNecessary()
```

Comment:
nit: Can we extract it as a preparatory commit

Reply:
What specifically? The commit is to "Avoid loading partitions eagerly in HiveSplitManager" and the changes to the `BackgroundHiveSplitLoader` are necessary to achieve that